File: sdks/java/src/dev/lunora/Key.java
package dev.lunora;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * The stable subscription key, ported from {@code shared/stable-key.ts}.
 *
 * <p>A key is compared verbatim against one produced by the reference TypeScript client, so every
 * spelling here must match ECMAScript exactly — a mismatch silently splits one subscription into
 * two.
 */
public final class Key {
    private Key() {}

    /**
     * Canonical JSON encoding of a pure-JSON tree: object keys sorted at every depth, arrays
     * keeping their order, null fields kept, {@link Wire#UNDEFINED} object fields dropped.
     *
     * <p>Runs on the OUTPUT of {@link Wire#encode}, so it only ever sees
     * null/Boolean/Double/String/List/Map.
     */
    public static String stableStringify(Object value) {
        StringBuilder out = new StringBuilder();

        writeStable(out, value);

        return out.toString();
    }

    /** The stable cache/dedup key for {@code value}. */
    public static String stableWireKey(Object value) {
        return stableStringify(Wire.encode(value));
    }

    private static void writeStable(StringBuilder out, Object value) {
        if (value == null || value == Wire.UNDEFINED) {
            out.append("null");
        } else if (value instanceof Boolean bool) {
            out.append(bool ? "true" : "false");
        } else if (value instanceof Number number) {
            out.append(formatNumber(number.doubleValue()));
        } else if (value instanceof String text) {
            out.append(jsonString(text));
        } else if (value instanceof List<?> items) {
            out.append('[');

            for (int index = 0; index < items.size(); index++) {
                if (index > 0) {
                    out.append(',');
                }

                writeStable(out, items.get(index));
            }

            out.append(']');
        } else if (value instanceof Map<?, ?> fields) {
            writeStableObject(out, fields);
        } else {
            out.append("null");
        }
    }

    private static void writeStableObject(StringBuilder out, Map<?, ?> fields) {
        List<String> keys = new ArrayList<>();

        for (Map.Entry<?, ?> entry : fields.entrySet()) {
            if (entry.getValue() == Wire.UNDEFINED) {
                continue;
            }

            keys.add(String.valueOf(entry.getKey()));
        }

        // Java's String.compareTo already compares UTF-16 code units, which is
        // exactly JavaScript's ordering — the one place the JVM needs no
        // adjustment where Go, Ruby, Rust and Swift all did.
        keys.sort(String::compareTo);
        out.append('{');

        for (int index = 0; index < keys.size(); index++) {
            if (index > 0) {
                out.append(',');
            }

            String key = keys.get(index);

            out.append(jsonString(key));
            out.append(':');
            writeStable(out, fields.get(key));
        }

        out.append('}');
    }

    /**
     * Renders a double exactly as {@code String(v)} does in JavaScript, which is what {@code
     * JSON.stringify} emits for a finite number.
     *
     * <p>Java's {@code Double.toString} always keeps a ".0" on integral values and switches to
     * exponent form outside 1e-3..1e7 spelled "1.0E-5"; ECMAScript drops the decimal, stays
     * positional between 1e-7 and 1e21, and writes "1e-7". The spellings must match or the key
     * differs.
     */
    static String formatNumber(double value) {
        if (Double.isNaN(value) || Double.isInfinite(value)) {
            return "null";
        }

        // Negative zero keeps its sign in a key — stableStringify emits the bare
        // token "-0" — and every integer conversion below drops it.
        if (value == 0.0 && 1.0 / value < 0.0) {
            return "-0";
        }

        if (value == Math.rint(value) && Math.abs(value) < 1e21) {
            return BigDecimal.valueOf(value)
                    .setScale(0, java.math.RoundingMode.HALF_UP)
                    .toBigInteger()
                    .toString();
        }

        double magnitude = Math.abs(value);

        if (magnitude >= 1e-6 && magnitude < 1e21) {
            return trimTrailingZeros(new BigDecimal(Double.toString(value)).toPlainString());
        }

        return exponential(value);
    }

    private static String exponential(double value) {
        for (int precision = 0; precision <= 17; precision++) {
            // Locale.ROOT is load-bearing, not tidiness: the default locale
            // decides the decimal separator, so on a German-locale JVM this
            // formats 1.5e-7 as "1,5e-07" — which then fails to parse back and,
            // worse, would produce a different subscription key than the same
            // code on an English-locale machine.
            String candidate = String.format(Locale.ROOT, "%." + precision + "e", value);

            if (Double.parseDouble(candidate) == value) {
                return normaliseExponent(candidate);
            }
        }

        return normaliseExponent(String.format(Locale.ROOT, "%.17e", value));
    }

    /** "1.000000e-07" -&gt; "1e-7": ECMAScript pads neither the mantissa nor the exponent. */
    private static String normaliseExponent(String text) {
        int marker = text.indexOf('e');

        if (marker < 0) {
            return text;
        }

        String mantissa = trimTrailingZeros(text.substring(0, marker));
        String exponent = text.substring(marker + 1);
        String sign = exponent.startsWith("-") ? "-" : "+";

        if (exponent.startsWith("-") || exponent.startsWith("+")) {
            exponent = exponent.substring(1);
        }

        exponent = exponent.replaceFirst("^0+(?=\\d)", "");

        return mantissa + "e" + sign + exponent;
    }

    private static String trimTrailingZeros(String text) {
        if (!text.contains(".")) {
            return text;
        }

        String trimmed = text.replaceFirst("0+$", "");

        return trimmed.endsWith(".") ? trimmed.substring(0, trimmed.length() - 1) : trimmed;
    }

    /**
     * Quotes a string the way {@code JSON.stringify} does: {@code <}, {@code >}, {@code &}, U+2028
     * and U+2029 all stay raw.
     */
    static String jsonString(String value) {
        StringBuilder out = new StringBuilder(value.length() + 2);

        out.append('"');

        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);

            switch (character) {
                case '"' -> out.append("\\\"");
                case '\\' -> out.append("\\\\");
                case '\n' -> out.append("\\n");
                case '\r' -> out.append("\\r");
                case '\t' -> out.append("\\t");
                case '\b' -> out.append("\\b");
                case '\f' -> out.append("\\f");
                default -> {
                    if (character < 0x20) {
                        out.append(String.format(Locale.ROOT, "\\u%04x", (int) character));
                    } else {
                        out.append(character);
                    }
                }
            }
        }

        out.append('"');

        return out.toString();
    }
}

File: sdks/swift/Sources/Lunora/Key.swift
import Foundation

extension Wire {
    /// Canonical JSON encoding of a pure-JSON tree: object keys sorted at every
    /// depth, arrays keeping their order, null fields kept, `WireUndefined`
    /// object fields dropped.
    ///
    /// Runs on the OUTPUT of ``encode(_:depth:)``, so it only ever sees
    /// null/bool/number/string/array/dictionary. Two argument records differing
    /// only in key insertion order collapse to one key — which is the point:
    /// this de-duplicates subscriptions, and it is compared verbatim against a
    /// key produced by the reference TypeScript client.
    ///
    /// It is also the transport's JSON WRITER for RPC and batch bodies: it cannot
    /// fail, and it spells numbers and strings exactly as `JSON.stringify` does.
    public static func stableStringify(_ value: Any?) -> String {
        switch value {
        case nil, is NSNull: return "null"
        case is WireUndefined: return "null"
        default: break
        }

        if let number = value as? NSNumber {
            if CFGetTypeID(number) == CFBooleanGetTypeID() { return number.boolValue ? "true" : "false" }
            return formatNumber(number)
        }

        if let boolean = value as? Bool { return boolean ? "true" : "false" }
        if let string = value as? String { return jsonString(string) }
        if let double = value as? Double { return formatDouble(double) }
        if let array = value as? [Any] { return "[" + array.map { stableStringify($0) }.joined(separator: ",") + "]" }
        if let dictionary = value as? [String: Any] { return stableObject(dictionary) }

        return "null"
    }

    /// The stable cache/dedup key for `value`.
    public static func stableWireKey(_ value: Any?) throws -> String {
        stableStringify(try encode(value))
    }

    private static func stableObject(_ value: [String: Any]) -> String {
        // JavaScript compares strings by UTF-16 code unit. Swift's `<` compares
        // by Unicode scalar, which agrees inside the BMP but not above it: an
        // astral character is its high surrogate (0xD83D) as UTF-16 yet its
        // full scalar (0x1F600) to Swift, so it sorts before U+FFFD there and
        // after it here. Comparing UTF-16 views reproduces JavaScript exactly.
        let pairs = value.filter { !($0.value is WireUndefined) }
        let sorted = pairs.sorted { lessUTF16($0.key, $1.key) }
        return "{" + sorted.map { "\(jsonString($0.key)):\(stableStringify($0.value))" }.joined(separator: ",") + "}"
    }

    private static func lessUTF16(_ a: String, _ b: String) -> Bool {
        var left = utf16Units(a).makeIterator()
        var right = utf16Units(b).makeIterator()
        while true {
            switch (left.next(), right.next()) {
            case (let l?, let r?):
                if l != r { return l < r }
            case (nil, .some): return true
            case (.some, nil): return false
            case (nil, nil): return false
            }
        }
    }

    private static func formatNumber(_ number: NSNumber) -> String {
        let type = CFNumberGetType(number as CFNumber)
        switch type {
        case .float32Type, .float64Type, .floatType, .doubleType, .cgFloatType:
            return formatDouble(number.doubleValue)
        default:
            return number.stringValue
        }
    }

    /// Renders a double exactly as `String(v)` does in JavaScript, which is what
    /// `JSON.stringify` emits for a finite number (ECMA-262 Number::toString).
    ///
    /// Swift's `description` already finds the SHORTEST digit string that reads
    /// back as the same double; only its layout differs ("1e-05", "1.0",
    /// "1e+16"). So the digits `d1…dk` and the exponent `n` (value = 0.d1…dk ×
    /// 10^n) are taken from it and laid out the ECMAScript way. A fixed-precision
    /// search spelled three adjacent doubles near -6e-6 identically, so a
    /// subscription keyed on one received another's frames.
    static func formatDouble(_ value: Double) -> String {
        if value.isNaN || value.isInfinite { return "null" }
        if value == 0 { return value.sign == .minus ? "-0" : "0" }

        let sign = value < 0 ? "-" : ""
        let text = abs(value).description
        let parts = text.split(separator: "e", maxSplits: 1)
        let mantissa = parts[0].split(separator: ".", maxSplits: 1, omittingEmptySubsequences: false)
        let whole = String(mantissa[0])
        var digits = whole + (mantissa.count > 1 ? String(mantissa[1]) : "")
        var n = whole.count + (parts.count > 1 ? Int(parts[1]) ?? 0 : 0)

        while digits.hasPrefix("0") {
            digits.removeFirst()
            n -= 1
        }

        while digits.hasSuffix("0") { digits.removeLast() }

        let k = digits.count

        if k <= n, n <= 21 { return sign + digits + String(repeating: "0", count: n - k) }
        if 0 < n, n <= 21 { return sign + digits.prefix(n) + "." + digits.dropFirst(n) }
        if -6 < n, n <= 0 { return sign + "0." + String(repeating: "0", count: -n) + digits }

        let exponent = n - 1
        let fraction = k > 1 ? "." + digits.dropFirst() : ""

        return sign + digits.prefix(1) + fraction + "e" + (exponent >= 0 ? "+" : "-") + String(abs(exponent))
    }

    /// A string's UTF-16 code units as its backing `NSString` holds them.
    ///
    /// Not `value.utf16`: a Swift view repairs a lone surrogate to U+FFFD, while
    /// the bridged `NSString` a truncation like `(s as NSString).substring(to:)`
    /// returns still holds it — and that unit is what `JSON.stringify` writes and
    /// what JavaScript sorts by.
    static func utf16Units(_ value: String) -> [UInt16] {
        let string = value as NSString
        var units = [UInt16](repeating: 0, count: string.length)

        string.getCharacters(&units, range: NSRange(location: 0, length: string.length))

        return units
    }

    /// Quotes a string the way `JSON.stringify` does: `"`, `\\`, the control
    /// characters, and a LONE surrogate as `\udXXX` (lowercase hex). `<`, `>`,
    /// `&`, U+2028 and U+2029 stay raw.
    ///
    /// This is also the transport's string writer, so it must never fail: a lone
    /// surrogate made `JSONSerialization` throw, and the offline queue re-queued
    /// that write — and every one behind it — on every flush, forever.
    static func jsonString(_ value: String) -> String {
        let units = utf16Units(value)
        var quoted = "\""
        var index = 0

        while index < units.count {
            let unit = units[index]
            index += 1

            switch unit {
            case 0x22: quoted += "\\\""
            case 0x5C: quoted += "\\\\"
            case 0x0A: quoted += "\\n"
            case 0x0D: quoted += "\\r"
            case 0x09: quoted += "\\t"
            case 0x08: quoted += "\\b"
            case 0x0C: quoted += "\\f"
            case 0xD800...0xDBFF where index < units.count && (0xDC00...0xDFFF).contains(units[index]):
                let low = units[index]
                index += 1
                quoted.unicodeScalars.append(Unicode.Scalar(0x10000 + (UInt32(unit - 0xD800) << 10) + UInt32(low - 0xDC00))!)
            case 0xD800...0xDFFF:
                quoted += String(format: "\\u%04x", unit)
            default:
                if unit < 0x20 {
                    quoted += String(format: "\\u%04x", unit)
                } else {
                    quoted.unicodeScalars.append(Unicode.Scalar(unit)!)
                }
            }
        }

        return quoted + "\""
    }
}
